File: backend/src/models/transaction.rs
// models/transaction.rs

use serde::{self, Serialize, Deserialize};
use sqlx::prelude::Type;

#[derive(Debug, Serialize, Deserialize, Type, Clone, Copy, PartialEq, Eq)]
#[sqlx(type_name = "CHAR(3)")]
#[serde(rename_all = "UPPERCASE")]
pub enum PaymentMode {
    CSH,    // Cash
    CCD,    // Credit Card
    DCD,    // Debit Card
    NBK,    // Net Banking
    UPI,    // Unified Payments Interface
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Transaction {
    pub txn_id: i64,
    pub total_amount: Option<f32>,
    pub txn_status: Option<String>,  // 'PENDING', 'COMPLETE', 'FAILED'
    pub payment_mode: Option<String>, // 'UPI', 'CARD', 'CASH', 'NETBANKING'
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CreateTransaction {
    pub total_amount: f32,
    pub txn_status: String,  // 'PENDING', 'COMPLETE', 'FAILED'
    pub payment_mode: PaymentMode,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct UpdateTransactionStatus {
    pub txn_id: i64,
    pub txn_status: Option<String>,  // 'PENDING', 'COMPLETE', 'FAILED'
}

#[derive(Debug, Deserialize)]
pub struct CancelBookingRequest {
    pub booking_id: i64,
    pub refund_amount: f64,
    pub txn_id: i64,
}

File: frontend/lib/data/models/transaction.dart
class PaymentTransaction {
  final int txnId;
  final double totalAmount;
  final String txnStatus;
  final String paymentMode;

  PaymentTransaction({
    required this.txnId,
    required this.totalAmount,
    required this.txnStatus,
    required this.paymentMode,
  });

  Map<String, dynamic> toJson() {
    return {
      'txn_id': txnId,
      'total_amount': totalAmount,
      'txn_status': txnStatus,
      'payment_mode': paymentMode,
    };
  }

  factory PaymentTransaction.fromJson(Map<String, dynamic> json) {
    return PaymentTransaction(
      txnId: json['txn_id'],
      totalAmount: json['total_amount'],
      txnStatus: json['txn_status'],
      paymentMode: json['payment_mode'],
    );
  }
}

class CreateTransaction {
  final double totalAmount;
  final String txnStatus;
  final String paymentMode;

  CreateTransaction({
    required this.totalAmount,
    required this.txnStatus,
    required this.paymentMode,
  });

  Map<String, dynamic> toJson() {
    return {
      'total_amount': totalAmount,
      'txn_status': txnStatus,
      'payment_mode': paymentMode,
    };
  }

  factory CreateTransaction.fromJson(Map<String, dynamic> json) {
    return CreateTransaction(
      totalAmount: json['total_amount'],
      txnStatus: json['txn_status'],
      paymentMode: json['payment_mode'],
    );
  }
}

